applications/flash_attention_v2/kernel/xe_flash_attn_gemm.hpp

```diff
@@ -317,28 +333,38 @@ class GemmUniversalAttention {
       // MAIN LOOP: loop over K and V, perform fused attention + online softmax
       for (int nblock = 0; nblock < nblock_limit - static_cast<int>(CausalMask); nblock++) {
         barrier_arrive(barrier_scope);
-        // 1) Load K (performed inside mmaQK)
+
+        bool is_KV_cache = (nblock < nblock_cache);
+
+        // 1) Load KV (performed inside mmaQK)
+        auto gK_ = is_KV_cache ? gK_cache(_, _, nblock, _) : gK(_, _, nblock - nblock_cache, _);
+        auto gV_ = is_KV_cache ? gV_cache(_, _, nblock) : gV(_, _, nblock - nblock_cache);
+
         // 2) Create Tensor S
         Tensor tSr = make_tensor<ElementAccumulator>(Shape<Int<Vec>, Int<FragsM>, Int<FragsN>>{});
         clear(tSr);
 
         // 3) Perform GEMM S = Q*K
-        collective_mma.mmaQK(tSr, gQ, gK(_, _, nblock, _), tSr, ceil_div(head_size_qk, QK_BLK_K), mainloop_params);
+        collective_mma.mmaQK(tSr, gQ, gK_, tSr, ceil_div(head_size_qk, QK_BLK_K), mainloop_params, is_KV_cache);
 
         // we only need one block ahead, there is enough gap to prefetch it while doing softmax. because the gap between the two MMA is big,
         // prefetching it the same way as cutlass K matrix does not make sense
-        prefetch(tiled_prefetch_v, pVgV(_, _, _ , nblock));
+        (is_KV_cache) ? prefetch(tiled_prefetch_v.with(mainloop_params.gmem_tensor_v_cache), pVgV_cache(_, _, _ , nblock))
+                      : prefetch(tiled_prefetch_v, pVgV(_, _, _ , nblock - nblock_cache));
 
+        // 4) Fused softmax
         CollectiveSoftmaxEpilogue softmax(params.softmax);
         softmax(nblock == 0, tSr, max_reg, sum_reg, out_reg);
 
-        collective_mma.mmaPV(out_reg, tSr, gV(_, _ , nblock), out_reg, mainloop_params);
+        // 5) Perform GEMM O = S*V
+        collective_mma.mmaPV(out_reg, tSr, gV_, out_reg, mainloop_params, is_KV_cache);
         
         // Prefetch the next K tile
         // there is no need to gaurd it with if statememt as prefetch will ignore out of bound reading
+        prefetch_K = (nblock + DispatchPolicy::Stages < nblock_cache) ? Pefetch_K{tiled_prefetch_k.with(mainloop_params.gmem_tensor_k_cache)}: tiled_prefetch_k;
         CUTLASS_PRAGMA_UNROLL
         for (int j = 0; j < size<4>(pKgK); j++) {
-          prefetch(tiled_prefetch_k, pKgK(_, _, _, nblock + DispatchPolicy::Stages, j));
+         prefetch(prefetch_K, pKgK_cache(_, _, _, nblock + DispatchPolicy::Stages, j));
```

Comment:
Should the prefetch be done on `pKgK_cache` regardless of `is_KV_cache`, or should it be conditionally still on `pKgK`?

Reply:
Yes I will fix it

Reply:
Fixed